### DISCUSSION TITLE
Should empty strings not add attributes?

### DISCUSSION THREAD

**Discussion by SRM256** [GridProtectionAlliance/gsf#85] [Ideas]
Looking at ContainerRecord specifically, but the same situation exists ubiquitously... in properties such as Notes, the "set" property unconditionally writes the value. If the value is an empty string, should the ContainerRecord remove any existing entry in the vector instead of adding a blank entry (as it does today)? The spec doesn't say one way or the other, but it seems like a better policy.

**Comment by StephenCWills** [comment]
The main reason I have against the policy of special-casing the empty string has to do with consistency in the API. Not every C# data type used by the library has a value that can represent emptiness. This means that in some cases, the `RemoveElement()` method is the only way certain elements can be removed from a container. Therefore, we can separate the duties of each of the operations on the elements: the property getter retrieves the value, the property setter sets the value, and the `RemoveElement()` method removes the element. This gives the user predictably defined ways of interacting with the file without having to consider semantic special cases dependent on the C# data type used to represent the value of a PQDIF element. As an added bonus, the library doesn't restrict the user from adding an empty string to the PQDIF file, if that seems at all useful for any use case.

Having said all that, I can also think of some arguments for special-casing the empty string as well so I'm certainly willing to reconsider my stance on the matter. May I ask what sparked this idea? Did you have a use case in mind for this change, or does it simply seem more intuitive to you?

**Reply by SRM256** [reply]
It wasn't so much that empty string is special from data type side... every property does have a default value, and when setting to the default value, we could not set or wipe out the property. I've got some files that are already large, and I don't want to make them even larger, so I was writing a bunch of code to say "if the input is the default, don't bother writing it". I found myself wishing that "don't write if default" was just the normal operation of the properties so that we didn't add unnecessary tags.
